sprint 1/ step 2

playboifufalo opened the pull request:
Добавлены модальные окна, оверлей для них, а также стили для отображения окон для ингредиентов и заказов.

В индексе добавлен айди для корректного открытия модальных окон, в аппе поменялся стиль на <main/>, а также добавлены файлы отдельно для отображения заказа и информации об ингредиенте.

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

К сожалению, работу придется отклонить без потери итерации

* В пулл-реквесте должен быть код изменённых файлов в `Files changed`, где я бы мог оставлять комментарии. Вам нужно делать ветку и не трогать ее до тех пор, пока работа не будет принята (решена) после ревью. Только тогда нужно мерджить ветку в `main`. Сейчас у Вас в файлах просто пробелы сделаны, чтобы 12 файлов отображалось в `Files changed`, но там нет кода проекта целиком


Еще заметил ошибки:


* Вам нужно удалить 3 лишних файла https://disk.yandex.ru/i/7lkSK2hAn_C_tw 
* Вам нужно использовать стили и дизайн по макету в фигме https://disk.yandex.ru/i/TUkimbUwCn-DLQ 
*  Попапы **не должны закрываться** при клике **внутри** них, так как внутри могут быть формы, текст, картинки, на которые пользователь будет нажимать, чтобы скопировать и т.д.
* Верхняя и нижняя булки должны **отличаться** по форме. Это делается с помощью атрибутов `type`
* Между булками должны быть разные ингредиенты, у которых должна быть иконка `DragIcon` и иконка `Мусорный бачок`
* Все стили по заданию  описываются в модульных `.module.css` файлах, а в коде компонентов используются модульные `css`-стили.
* Папки с компонентами (`burgerConstructor`, `ingredientsDetails`, `modals`) нужно называть так, как сам компонент (с большой буквы), чтобы соблюсти единый формат наименования, а то у вас смешанный стиль
*

playboifufalo committed:
Добавлены модальные окна, оверлей для них, а также стили для отображения окон для ингредиентов и заказов.
В индексе добавлен айди для корректного открытия модальных окон, в аппе поменялся стиль на <main/>, а также добавлены файлы отдельно для отображения заказа и информации об ингредиенте.

gennady-bars commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -1,8 +1,12 @@
+
     const ingredients = [
         {
             id: 1,
```
Моковые данные нужно вынести в `utils`, чтобы не засорять код компонента

gennady-bars commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -1,8 +1,12 @@
+import OrderDetails from '../order-details/order-details';
 
 const BurgerConstructor = () => {
+    const [isModalVisible, setIsModalVisible] = useState(false);
```
**Можно лучше**

Если интересно, можно создать **кастомный хук** для управления модальным окном:

```js
import { useState, useCallback } from "react";

// кастомные хуки всегда должны начинаться с глагола `use`, чтобы реакт понял, что это хук. Он следит за их вызовами
export const useModal = () => {
  const [isModalOpen, setIsModalOpen] = useState(false);

  // `useCallback` нужен для того, чтобы зафиксировать ссылку на функцию. Таким образом уменьшится кол-во перерисовок компонента, куда будет передана эта функция
  const openModal = useCallback(() => {
    setIsModalOpen(true);
  }, []);

  const closeModal = useCallback(() => {
    setIsModalOpen(false);
  }, []);

  return {
    isModalOpen,
    openModal,
    closeModal,
  };
};

```

Этот код помещают в отдельный файл `useModal.js` в папке `hooks` и импортируют функцию туда, где нужно создать показатели открытости/закрытости попапа (у каждого попапа они свои)

И теперь нужно просто вызывать хук в том компоненте, где нужно управлять попапом. Всего одна строчка кода будет (под каждый попап своя):

```js
  const { isModalOpen, openModal, closeModal } = useModal();

````

gennady-bars commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -1,14 +1,42 @@
+
+        document.addEventListener('keydown', handleEscClose);
+        return () => {
+            document.removeEventListener('keydown', handleEscClose);
```
Обработчик `Escape` нужно поместить внутрь `Modal`.  Это очистит код `компонента`  и будет единое место объявления обработчика. Там же будет правильно удаляться обработчик при размонтировании `Modal`

gennady-bars replied:
* Вы не удалили кое-что https://github.com/playboifufalo/react-burger/pull/4#discussion_r1740092343 


Обработчик `Escape` уже есть внутри `Modal`, поэтому нужно удалить из `BurgerIngredients`

gennady-bars commented on src/components/ingredient-details/ingredient-details.jsx:
```diff
@@ -0,0 +1,33 @@
+import React from 'react';
+import styles from './ingredient-details.module.css';
+
+const IngredientDetails = ({ ingredient }) => {
```
`IngredientDetails` забыли типизировать

gennady-bars replied:
Нужно посмотреть все компоненты и **типизировать** их `пропсы`, если где-то пропустили.

gennady-bars commented on src/components/modals/modal/modal.jsx:
```diff
@@ -0,0 +1,43 @@
+        document.addEventListener('keydown', handleEscClose);
+        return () => {
+            document.body.style.overflow = '';
+            document.removeEventListener('keydown', handleEscClose);
```
Правильно удаляется обработчик

gennady-bars commented on src/components/modals/modal-overlay/modal-overlay.jsx:
```diff
@@ -0,0 +1,11 @@
+
+const ModalOverlay = (onClose) => {
+    return (
+        <div className={styles.overlay} onClick={onClose}></div>
```
* Попапы должны еще закрываться по **оверлею**

playboifufalo replied:
Что означает закрываться по оверлею?

gennady-bars replied:
> Что означает закрываться по оверлею?

При клике на затемненную область вокруг открытого попапа. Она называется **оверле**, поэтому и компонент называется `ModalOverlay`

gennady-bars commented on src/components/modals/modal/modal.jsx:
```diff
@@ -0,0 +1,43 @@
+
+    return ReactDOM.createPortal(
+        (
+            <div className={styles.modal_overlay}>
```
`<div className={styles.modal_overlay}>`


Вот это должно быть внутри `ModalOverlay` (затемнение вокруг попапа, при клике на которое закрывается попап)

gennady-bars commented on src/components/app/app.jsx:
```diff
@@ -9,12 +9,13 @@ function App() {
   return (
```
Теперь в `App` нужно делать запрос на сервер и получать оттуда список **ингредиентов**. Моковые ингредиенты теперь не нужны в проекте

gennady-bars commented on src/components/order-details/order-details.jsx:
```diff
@@ -0,0 +1,20 @@
+
+const OrderDetails = () => {
+    return (
+        <div className={styles.orderContainer}>
```
* Попап заказа нужно сделать по макету в фигме

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на **гитхабе** все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная, но есть некоторые недочеты: 


* Теперь в `App` нужно делать запрос на сервер и получать оттуда список **ингредиентов**. Моковые ингредиенты теперь не нужны в проекте
* Попапы должны еще закрываться по **оверлею**
* Попап заказа нужно сделать по макету в фигме
* При открытии попапов в консоли ошибка https://disk.yandex.ru/i/roXtu2aLgWxiDw 
* Моковые данные нужно вынести в `utils`, чтобы не засорять код компонента
* Обработчик `Escape` нужно поместить внутрь `Modal`.  Это очистит код `компонента`  и будет единое место объявления обработчика. Там же будет правильно удаляться обработчик при размонтировании `Modal`
* Инлайн-стилей `style={{`не должно быть в проекте, если они не меняются **динамически** в коде компонента. Это затрудняет управление стилями. Все стили описываются в модульных `.module.css` файлах, а в коде компонентов используются модульные `css`-стили.
* Вы не исправили некоторые замечания с прошлого раза: https://github.com/playboifufalo/react-burger/pull/3#discussion_r1738438455 и https://github.com/playboifufalo/react-burger/pull/3#discussion_r1738439381 
* `IngredientDetails` забыли типизировать
* Нужно посмотреть все компоненты и **типизировать** их `пропсы`, если где-то пропустили.

playboifufalo committed:
Убрал моковые данные, добавил типы пропсов и

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

К сожалению, работу придется отклонить без потери итерации

* Вы не справили половину моих замечаний, начиная отсюда https://github.com/playboifufalo/react-burger/pull/4#pullrequestreview-2274366384 
* При открытии попапа заказа в консоли ошибка 
* Попапы не закрываются по оверлею

playboifufalo commented:
> Здравствуйте. (**Нужно развернуть общий комментарий ↓**)
> 
> К сожалению, работу придется отклонить без потери итерации
> 
> * Вы не справили половину моих замечаний, начиная отсюда [sprint 1/ step 2 #4 (review)](https://github.com/playboifufalo/react-burger/pull/4#pullrequestreview-2274366384)
> * При открытии попапа заказа в консоли ошибка
> * Попапы не закрываются по оверлею

Здравствуйте, разве я не изменил большую часть замечаний с прошлого раза? Мне кажется, что почти все было изменено

playboifufalo committed:
Изменены стили в соответствии с фигмой, изменено использование модальных окон

playboifufalo committed:
Добавлена иконка в ордер

gennady-bars commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -1,14 +1,74 @@
+        const fetchIngredients = async () => {
+            try {
+                const response = await fetch('https://norma.nomoreparties.space/api/ingredients');
+                const data = await response.json();
```
`                const data = await response.json();
`

Любой ответ от сервера нужно проверять на `ok` (на **корректность**) и вызывать ошибку, если ответ не `ok`. Таким образом, в `catch` Вы будете ловить и ошибки сервера, и ошибки невалидных данных.


```js
.then(res => {
                if (res.ok) {
                    return res.json();
                }
                return Promise.reject(`Ошибка ${res.status}`);
            })

```

gennady-bars replied:
* Вы кое-что не исправили https://github.com/playboifufalo/react-burger/pull/4#discussion_r1745562352

gennady-bars commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -1,14 +1,74 @@
+    if (error) {
+        return <div>{error}</div>;
+    }
 
```
* Инлайн-стилей `style={{`не должно быть в проекте, если они не меняются **динамически** в коде компонента. Это затрудняет управление стилями. Все стили описываются в модульных `.module.css` файлах, а в коде компонентов используются модульные `css`-стили. В `BurgerIngredients` нужно удалить `style={{` с 79й строчки

gennady-bars commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -75,8 +135,38 @@ const BurgerIngredients = () => {
+        proteins: PropTypes.number,
+        fat: PropTypes.number,
+        carbohydrates: PropTypes.number,
+    })),
```
Тип для ингредиента (propTypes) повторяется в некоторых компонентах. Не нужно его копировать. Это очень сложный объект. Нужно вынести его в константу `IngredientType` в отдельный файл `types.js` в папку `utils` и импортировать туда, где нужно

gennady-bars commented on src/components/ingredient-details/ingredient-details.jsx:
```diff
@@ -0,0 +1,48 @@
+        calories: PropTypes.number.isRequired,
+        proteins: PropTypes.number.isRequired,
+        fat: PropTypes.number.isRequired,
+        carbohydrates: PropTypes.number.isRequired,
```
Тип для ингредиента (propTypes) повторяется в некоторых компонентах. Не нужно его копировать. Это очень сложный объект. Нужно вынести его в константу `IngredientType` в отдельный файл `types.js` в папку `utils` и импортировать туда, где нужно

gennady-bars replied:
Нужно везде использовать `IngredientType` для типизации объекта ингредиента. Не нужно дублировать код

gennady-bars commented on src/components/modals/modal/modal.jsx:
```diff
@@ -0,0 +1,53 @@
+        };
+    }, [onClose]);
+    
+    const modalClass = type === 'order' ? styles.modalContent_order : styles.modalContent_ingr;
```
`    const modalClass = type === 'order' ? styles.modalContent_order : styles.modalContent_ingr;
`

Про заказ или ингредиент `Modal` не должен знать, ведь это универсальный компонент для любых попапов


Вам нужно проверять по `title`. Если есть `title`, то это попап ингредиента, иначе заказа


`modalContent_order` и `modalContent_ingr` тоже нужно переименовать нейтрально

```js

    const modalClass = title ? styles.withTitle : styles.noTitle;

```

gennady-bars reviewed (commented):
Еще немного и все будет отлично:


* При открытии попапа заказа в консоли ошибка https://disk.yandex.ru/i/qVtP7Y2FqS_Msg 
* Попапы по заданию должны закрываться при клике на **затемненную область вокруг попапа**, называемую в программировании **оверлей** https://disk.yandex.ru/i/-AD8c79H2XKh0g . Вы всегда можете спросить наставника о том, что вызывает трудности. Он поможет
* Вы не удалили кое-что https://github.com/playboifufalo/react-burger/pull/4#discussion_r1740092343  Обработчик `Escape` уже есть внутри `Modal`, поэтому нужно удалить из `BurgerIngredients`
* Любой ответ от сервера нужно проверять на `ok` (на **корректность**) и вызывать ошибку, если ответ не `ok`. Таким образом, в `catch` Вы будете ловить и ошибки сервера, и ошибки невалидных данных.
* Инлайн-стилей `style={{`не должно быть в проекте, если они не меняются **динамически** в коде компонента. Это затрудняет управление стилями. Все стили описываются в модульных `.module.css` файлах, а в коде компонентов используются модульные `css`-стили. В `BurgerIngredients` нужно удалить `style={{` с 79й строчки
* Вы не справили замечание https://github.com/playboifufalo/react-burger/pull/4#discussion_r1740092905
* Тип для ингредиента (propTypes) повторяется в некоторых компонентах. Не нужно его копировать. Это очень сложный объект. Нужно вынести его в константу `IngredientType` в отдельный файл `types.js` в папку `utils` и импортировать туда, где нужно
* Про заказ или ингредиент `Modal` не должен знать, ведь это универсальный компонент для любых попапов

playboifufalo committed:
Теперь модальное окно работает при обозначении title, также теперь данные передаются через api. Моковые данные и проптипы убраны в types.js

playboifufalo committed:
Изменена работа модальных окон, изменен их стиль

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

К сожалению, работу придется отклонить без потери итерации


* В консоли ошибка. При открытии попапа заказа ошибка
* Вы не исправили большую часть замечаний отсюда  https://github.com/playboifufalo/react-burger/pull/4#pullrequestreview-2283098957

playboifufalo committed:
Теперь в консоли не появляется ошибка, потому что исправлена работа проптайпсов. Ошибки теперь отображаются при фетче ингредиентов, а стиль не появляется, если ингредиенты взяты из API

playboifufalo commented:
> Здравствуйте. (**Нужно развернуть общий комментарий ↓**)
> 
> К сожалению, работу придется отклонить без потери итерации
> 
> * В консоли ошибка. При открытии попапа заказа ошибка
> * Вы не исправили большую часть замечаний отсюда  [sprint 1/ step 2 #4 (review)](https://github.com/playboifufalo/react-burger/pull/4#pullrequestreview-2283098957)

Здравствуйте! Остальные замечания исправил. В консоли теперь нету ошибки, проптайпсы есть там, где они нужны, а ненужные стили больше не появляются в коде (там, где они взяты из пропсов).
Также модальное окно и его стиль открывается в зависимости от тайтла компонента

gennady-bars commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -1,14 +1,62 @@
+        const fetchIngredients = async () => {
+            try {
+                const response = await fetch('https://norma.nomoreparties.space/api/ingredients');
+                const data = await response.json();
```
* Кое-что не исправили https://github.com/playboifufalo/react-burger/pull/4/files#r1745562352


Вот так должно быть тут

```js
        try {
                const response = await fetch('https://norma.nomoreparties.space/api/ingredients');
                if (!response.ok) {
                    return Promise.reject(`Ошибка ${response.status}`);
                }
                const data = await response.json();
            
                if (data.success) {
```

gennady-bars commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -29,32 +77,32 @@ const BurgerIngredients = () => {
                 </div>
```
* Нужно заменить `style={{ display: 'flex' }}` на стили из **модульного файла стилей: будет примерно так `styles.tabs`. Это в `burger-ingredients.jsx` на 67й строчке

gennady-bars reviewed (commented):
Еще немного и все будет отлично:


* Нужно везде использовать `IngredientType` для типизации объекта ингредиента. Не нужно дублировать код в `IngredientDetails`
* Нужно заменить `style={{ display: 'flex' }}` на стили из **модульного файла стилей: будет примерно так `styles.tabs`. Это в `burger-ingredients.jsx` на 67й строчке
* Кое-что не исправили https://github.com/playboifufalo/react-burger/pull/4/files#r1745562352

playboifufalo committed:
Теперь все должно быть отлично :)

gennady-bars commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -33,14 +30,10 @@ const BurgerIngredients = () => {
+                    return Promise.reject(`Ошибка ${response.status}`);
                 }
-            } catch (error) {
-                setError('Failed to load ingredients');
```
Блок `catch` не нужно было удалять. Он ловит те ошибки, который с сервера приходят или те, что Вы сами выкидываете ранее, если ответ от сервера не `ok`

gennady-bars commented on src/components/ingredient-details/ingredient-details.jsx:
```diff
@@ -34,14 +34,6 @@ const IngredientDetails = ({ ingredient }) => {
-        fat: PropTypes.number.isRequired,
-        carbohydrates: PropTypes.number.isRequired,
-    }).isRequired,
 };
```
В `IngredientDetails` нужна типизация `ingredient`


```js

IngredientDetails.propTypes = {
  ingredient: IngredientType,
};
```

gennady-bars commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -33,14 +30,10 @@ const BurgerIngredients = () => {
                 const response = await fetch('https://norma.nomoreparties.space/api/ingredients');
-                const data = await response.json();
-                if (data.success) {
-                    setIngredients(data.data);
```
* Нет ингредиентов на сайте. Вам не нужно было менять логику с `setIngredients`.


Вот такой код получится

```js

  useEffect(() => {
    const fetchIngredients = async () => {
      try {
        const response = await fetch(
          "https://norma.nomoreparties.space/api/ingredients"
        );
        if (!response.ok) {
          return Promise.reject(`Ошибка ${response.status}`);
        }
        const data = await response.json();
        setIngredients(data.data);
      } catch (error) {
        setError("Failed to load ingredients");
      } finally {
        setIsLoading(false);
      }
    };

    fetchIngredients();
  }, []);

```

gennady-bars reviewed (commented):
Кое-что нужно подправить:

* Нет ингредиентов на сайте. Вам не нужно было менять логику с `setIngredients`.
* Блок `catch` не нужно было удалять. Он ловит те ошибки, который с сервера приходят или те, что Вы сами выкидываете ранее, если ответ от сервера не `ok`
* В `IngredientDetails` нужна типизация `ingredient`

playboifufalo committed:
Теперь все должно быть отлично

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
